File: cpp/lib/testcontainers/core/CgroupnsMode.cpp
#include <rust/cxx.h>
#include <rust_tc_bridge/lib.h>

#include "testcontainers/core/CgroupnsMode.hpp"

#include "details/BoxHelper.hpp"

namespace testcontainers {

CgroupnsMode::CgroupnsMode(RsCgroupnsMode *mode) noexcept
    : rimpl_(mode,
             [](RsCgroupnsMode *m) { ::rs_cgroupns_mode_destroy(details::box_from_raw(m)); }) {}

CgroupnsMode::CgroupnsMode(CgroupnsMode &&other) noexcept = default;

CgroupnsMode &CgroupnsMode::operator=(CgroupnsMode &&other) noexcept = default;

CgroupnsMode::~CgroupnsMode() noexcept = default;

CgroupnsMode CgroupnsMode::Private() noexcept {
  return CgroupnsMode(::rs_cgroupns_mode_private().into_raw());
}

CgroupnsMode CgroupnsMode::Host() noexcept {
  return CgroupnsMode(::rs_cgroupns_mode_host().into_raw());
}

bool CgroupnsMode::is_valid() const noexcept { return static_cast<bool>(rimpl_); }

} // namespace testcontainers

File: rust/src/core/cgroupns_mode.rs
use testcontainers::core::CgroupnsMode;

pub struct RsCgroupnsMode {
	mode: CgroupnsMode,
}

pub fn rs_cgroupns_mode_private() -> Box<RsCgroupnsMode> {
	Box::new(RsCgroupnsMode::new(CgroupnsMode::Private))
}

pub fn rs_cgroupns_mode_host() -> Box<RsCgroupnsMode> {
	Box::new(RsCgroupnsMode::new(CgroupnsMode::Host))
}

pub fn rs_cgroupns_mode_destroy(mode: Box<RsCgroupnsMode>) {
	drop(mode);
}

impl RsCgroupnsMode {
	pub fn new(mode: CgroupnsMode) -> Self {
		Self { mode }
	}
}

impl Into<CgroupnsMode> for RsCgroupnsMode {
	fn into(self) -> CgroupnsMode {
		self.mode
	}
}

